plaidml/edsl/edsl.h

```diff
@@ -1150,16 +1252,60 @@ inline Tensor Tensor::operator-() const { return intrinsicCall("neg", {*this});
 inline Tensor Tensor::operator~() const { return intrinsicCall("bit_not", {*this}); }
 inline Tensor Tensor::operator!() const { return intrinsicCall("logical_not", {*this}); }
 
-#define PLAIDML_EDSL_DEFINE_TENSOR_BINARY_OPS(_op_, _fn_)                                                 \
-  inline Tensor operator _op_(const Tensor& lhs, const Tensor& rhs) { return intrinsic(_fn_, lhs, rhs); } \
-  inline Tensor operator _op_(const Tensor& lhs, int rhs) { return intrinsic(_fn_, lhs, rhs); }           \
-  inline Tensor operator _op_(const Tensor& lhs, int64_t rhs) { return intrinsic(_fn_, lhs, rhs); }       \
-  inline Tensor operator _op_(const Tensor& lhs, uint64_t rhs) { return intrinsic(_fn_, lhs, rhs); }      \
-  inline Tensor operator _op_(const Tensor& lhs, double rhs) { return intrinsic(_fn_, lhs, rhs); }        \
-  inline Tensor operator _op_(int lhs, const Tensor& rhs) { return intrinsic(_fn_, lhs, rhs); }           \
-  inline Tensor operator _op_(int64_t lhs, const Tensor& rhs) { return intrinsic(_fn_, lhs, rhs); }       \
-  inline Tensor operator _op_(uint64_t lhs, const Tensor& rhs) { return intrinsic(_fn_, lhs, rhs); }      \
-  inline Tensor operator _op_(double lhs, const Tensor& rhs) { return intrinsic(_fn_, lhs, rhs); }
+template <typename T>
+class has_tensor {
+  typedef char one;
+  typedef int two;
+
+  template <typename C>
+  static one test(decltype(&C::operator Tensor));
+  template <typename C>
+  static two test(...);
+
+ public:
+  enum { value = sizeof(test<T>(0)) == sizeof(char) };
+};
+
+template <bool B, class T = void>
+using enable_if_t = typename std::enable_if<B, T>::type;
+
+struct LocatedTensor {
+  Tensor tensor;
+  edsl_source_location loc;
+
+  template <typename T, typename = enable_if_t<has_tensor<T>::value>>
+  LocatedTensor(T tensor, edsl_source_location loc = edsl_source_location::current())  // NOLINT
+      : tensor(tensor), loc(loc) {}
+
+  LocatedTensor(Tensor tensor, edsl_source_location loc = edsl_source_location::current())  // NOLINT
+      : tensor(tensor), loc(loc) {}
+};
+
+#ifdef _WIN32
```

Comment:
I think perhaps this should be `#ifdef MSVC`? or whatever the appropriate define to use for detecting MSVC instead of WIN32 (imagine if someone was trying to compile the EDSL using gcc-mingw32 on windows).